pkg/controller/remotecluster/controller.go

```diff
@@ -0,0 +1,250 @@
+		remoteVtepSynced:          remoteSubnetInformer.Informer().HasSynced,
+		localClusterNetworkLister: localClusterNetworkInformer.Lister(),
+		localClusterNetworkSynced: localClusterNetworkInformer.Informer().HasSynced,
+		remoteClusterQueue:        workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerName),
+		rcMgrQueue:                workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "remoteclustermanager"),
+		recorder:                  recorder,
+	}
+
+	remoteClusterInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: c.filterRemoteCluster,
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addOrDelRemoteCluster,
+			UpdateFunc: c.updateRemoteCluster,
+			DeleteFunc: c.addOrDelRemoteCluster,
+		},
+	})
+
+	return c
+}
+
+func (c *Controller) Run(stopCh <-chan struct{}) error {
+	defer runtimeutil.HandleCrash()
+	defer c.rcMgrQueue.ShutDown()
+	defer c.remoteClusterQueue.ShutDown()
+
+	klog.Infof("Starting %s controller", ControllerName)
+
+	klog.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, c.remoteClusterSynced, c.remoteSubnetSynced, c.remoteVtepSynced, c.localClusterSubnetSynced); !ok {
+		return fmt.Errorf("%s failed to wait for caches to sync", ControllerName)
+	}
+
+	// start workers
+	klog.Info("Starting workers")
+	go wait.Until(c.runRemoteClusterWorker, time.Second, stopCh)
+	go wait.Until(c.processRCManagerQueue, time.Second, stopCh)
+	go wait.Until(c.runOverlayNetIDWorker, time.Minute, stopCh)
+	go wait.Until(c.updateRemoteClusterStatus, HealthCheckPeriod, stopCh)
+	<-stopCh
+
+	c.closeRemoteClusterManager()
+
+	klog.Info("Shutting down workers")
+	return nil
+}
+
+func (c *Controller) closeRemoteClusterManager() {
+	// no need to lock
+	for _, mgr := range c.rcMgrCache.rcMgrMap {
+		mgr.Close()
+	}
+}
+
+func (c *Controller) runOverlayNetIDWorker() {
+	c.overlayNetIDMU.Lock()
+	defer c.overlayNetIDMU.Unlock()
+
+	networks, err := c.localClusterNetworkLister.List(labels.Everything())
+	if err != nil {
+		klog.Warningf("Can't list local cluster network. err=%v", err)
+		return
+	}
+	for _, network := range networks {
+		if network.Spec.Type == networkingv1.NetworkTypeOverlay {
+			n := network.DeepCopy()
+			c.OverlayNetID = n.Spec.NetID
+			break
+		}
+	}
+}
+
+// health checking and resync cache. remote cluster is managed by admin, it can be
+// treated as desired states
+func (c *Controller) updateRemoteClusterStatus() {
+	remoteClusters, err := c.remoteClusterLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Can't list remote cluster. err=%v", err)
+		return
+	}
+
+	var (
+		wg  sync.WaitGroup
+		cnt = 0
+	)
+	for _, rc := range remoteClusters {
+		r := rc.DeepCopy()
+		manager, exists := c.rcMgrCache.Get(r.Name)
+		if !exists {
+			continue
+		}
+		cnt = cnt + 1
+		wg.Add(1)
+		go c.updateSingleRCStatus(manager, r, &wg)
```

Comment:
use inline function here to avoid passing `wg` through functional parameters

Reply:
why? It is pointer

Reply:
do you know inline function?

Reply:
yes, and then?